Custom model flavor example (#7867)
mlflow/mlflow: docs/source/models.rst

    from mlflow.utils.environment import (
        _CONDA_ENV_FILE_NAME,
        _CONSTRAINTS_FILE_NAME,
        _PYTHON_ENV_FILE_NAME,
        _REQUIREMENTS_FILE_NAME,
        _mlflow_conda_env,
        _process_conda_env,
        _process_pip_requirements,
        _PythonEnv,
        _validate_env_arguments,
    )
    from mlflow.utils.file_utils import write_to
    from mlflow.utils.model_utils import (
        _add_code_from_conf_to_system_path,
        _get_flavor_configuration,
        _validate_and_copy_code_paths,
        _validate_and_prepare_target_save_path,
    )
    from mlflow.utils.requirements_utils import _get_pinned_requirement
    from sktime.utils.multiindex import flatten_multiindex

    _logger = logging.getLogger(__name__)

We continue by defining a set of important variables used throughout the code that follows.

The flavor name needs to be provided for every custom flavor and should reflect the name of the
library to be supported. It is saved as part of the flavor-specific attributes to the ``MLmodel``
configuration file. This example also defines some ``sktime`` specific variables. For illustration

> purposes only a subset of the available predict methods to be exposed via the

Reviewer: ```suggestion
purposes, only a subset of the available predict methods to be exposed via the
```

``_SktimeModelWrapper`` class is included when loading the model in its ``python_function`` flavor
(additional methods could be added in a similar fashion). Additionaly, the model serialization
formats, namely ``pickle`` (default) and ``cloudpickle``, are defined. Note that both serialization
modules require using the same python environment (version) in whatever environment this model is

> used for inference to ensure that the model will load with appropriate version of

Reviewer: ```suggestion
used for inference to ensure that the model will load with the appropriate version of
```

``save_model()`` and ``log_model()`` functions. In the ``save_model()`` function, the ``sktime``
model is saved to a specified output directory. Additionally, ``save_model()`` leverages the
:py:func:`mlflow.models.Model.add_flavor()` and :py:func:`mlflow.models.Model.save()` methods to
produce the ``MLmodel`` configuration containing the ``sktime`` and the ``python_function`` flavor.
The resulting configuration has several flavor-specific attributes, such as the flavor name and
``sktime_version``, which denotes the version of the ``sktime`` library that was used to train the
model.

The ``save_model()`` function also provides flexibility to add additional paramters which can be
added as flavor-specific attributes to the model configuration. In this example there is only one
flavor-specific parameter for specifying the model serialization format. All other paramters are
non-flavor specific (for a detailed description of these parameters take a look at
`mlflow.sklearn.save_model <https://mlflow.org/docs/latest/python_api/mlflow.sklearn.html#mlflow.sklearn.save_model>`_).
When creating your own custom flavor make sure to rename the ``sktime_model`` parameter in the

> ``save_model()`` function to reflect the flavor name to be supported.

Reviewer: ```suggestion
``save_model()`` function to reflect the name that you've chosen for your custom flavor.
```

> When creating your own custom flavor make sure to rename the ``sktime_model`` parameter in the
> ``save_model()`` function to reflect the flavor name to be supported.

Reviewer: ```suggestion
Note: When creating your own custom flavor, be sure rename the ``sktime_model`` parameter in both the
``save_model()`` and ``log_model()`` functions to reflect the name of your custom model flavor.
```

            )
        else:
            conda_env, pip_requirements, pip_constraints = _process_conda_env(conda_env)

        with open(os.path.join(path, _CONDA_ENV_FILE_NAME), "w") as f:
            yaml.safe_dump(conda_env, stream=f, default_flow_style=False)

        if pip_constraints:
            write_to(os.path.join(path, _CONSTRAINTS_FILE_NAME), "\n".join(pip_constraints))

        write_to(os.path.join(path, _REQUIREMENTS_FILE_NAME), "\n".join(pip_requirements))

        _PythonEnv.current().to_yaml(os.path.join(path, _PYTHON_ENV_FILE_NAME))

    def _save_model(model, path, serialization_format):
        with open(path, "wb") as out:
            if serialization_format == SERIALIZATION_FORMAT_PICKLE:
                pickle.dump(model, out)
            else:
                import cloudpickle

                cloudpickle.dump(model, out)

The ``save_model()`` function also writes the model dependencies to a ``requirements.txt`` and
``conda.yaml`` file in the model output directory. For this purpose the set of ``pip`` dependecies
produced by this flavor need to be added to the ``get_default_pip_requirements()`` function. In this

> example only the minimum required dependencies are provided but additional requirements needed for

Reviewer: ```suggestion
example only the minimum required dependencies are provided. In practice, additional requirements needed for
```

> preprocessing or post-processing steps could be included. Note that for any custom flavor the

Reviewer: ```suggestion
preprocessing or post-processing steps could be included. Note that for any custom flavor, the
```

:py:func:`mlflow.models.infer_pip_requirements()` method in the ``save_model()`` function will
return the default requirements defined in ``get_default_pip_requirements()`` as package imports are
only inferred for built-in flavors.

.. code-block:: python

    def get_default_pip_requirements(include_cloudpickle=False):
        pip_deps = [_get_pinned_requirement("sktime")]
        if include_cloudpickle:
            pip_deps += [_get_pinned_requirement("cloudpickle")]

        return pip_deps

    def get_default_conda_env(include_cloudpickle=False):
        return _mlflow_conda_env(
            additional_pip_deps=get_default_pip_requirements(include_cloudpickle)
        )

> Next we add the ``log_model()`` function which is just a wrapper around the

Reviewer: ```suggestion
Next, we add the ``log_model()`` function. This function is little more than a wrapper around the
```
